packages/ra-core/src/controller/field/ReferenceManyFieldBase.tsx

```diff
@@ -0,0 +1,143 @@
+ * `filter` prop for that.
+ *
+ * @example
+ * <ReferenceManyFieldBase filter={{ is_published: true }} reference="comments" target="post_id">
+ *    ...
+ * </ReferenceManyFieldBase>
+ */
+export const ReferenceManyFieldBase = <
+    RecordType extends RaRecord = RaRecord,
+    ReferenceRecordType extends RaRecord = RaRecord,
+>(
+    props: ReferenceManyFieldBaseProps<RecordType, ReferenceRecordType>
+) => {
+    const {
+        children,
+        debounce,
+        empty,
+        filter = defaultFilter,
+        page = 1,
+        pagination = null,
+        perPage = 25,
+        record,
+        reference,
+        resource,
+        sort = defaultSort,
+        source = 'id',
+        storeKey,
+        target,
+        queryOptions,
+    } = props;
+
+    const controllerProps = useReferenceManyFieldController<
+        RecordType,
+        ReferenceRecordType
+    >({
+        debounce,
+        filter,
+        page,
+        perPage,
+        record,
+        reference,
+        resource,
+        sort,
+        source,
+        storeKey,
+        target,
+        queryOptions,
+    });
+
+    if (
+        // there is an empty page component
+        empty &&
+        // there is no error
+        !controllerProps.error &&
+        // the list is not loading data for the first time
+        !controllerProps.isPending &&
+        // the API returned no data (using either normal or partial pagination)
+        (controllerProps.total === 0 ||
+            (controllerProps.total == null &&
+                // @ts-ignore FIXME total may be undefined when using partial pagination but the ListControllerResult type is wrong about it
+                controllerProps.hasPreviousPage === false &&
+                // @ts-ignore FIXME total may be undefined when using partial pagination but the ListControllerResult type is wrong about it
+                controllerProps.hasNextPage === false &&
+                // @ts-ignore FIXME total may be undefined when using partial pagination but the ListControllerResult type is wrong about it
+                controllerProps.data.length === 0)) &&
+        // the user didn't set any filters
+        !Object.keys(controllerProps.filterValues).length
+    ) {
+        return empty;
+    }
+
+    return (
+        <ResourceContextProvider value={reference}>
+            <ListContextProvider value={controllerProps}>
+                {children}
+                {pagination}
+            </ListContextProvider>
+        </ResourceContextProvider>
+    );
+};
+
+export interface ReferenceManyFieldBaseProps<
+    RecordType extends Record<string, any> = Record<string, any>,
+    ReferenceRecordType extends Record<string, any> = Record<string, any>,
+> extends UseReferenceManyFieldControllerParams<
+        RecordType,
+        ReferenceRecordType
+    > {
+    children: ReactNode;
+    empty?: ReactNode;
+    pagination?: ReactElement;
```

Comment:
```suggestion
    pagination?: ReactNode;
```
You can probably remove the `ReactElement` import as well